File: lib/payment_preview.rb
# frozen_string_literal: true

require 'json'
require_relative 'dues_calculator'

payload = JSON.parse($stdin.read)
calculator = Gvac::DuesCalculator
obligations = payload.fetch('obligations').map do |row|
  period = calculator.month(row.fetch('period_start'))
  Gvac::Obligation.new(
    period: period,
    amount_due_ngn: Integer(row.fetch('amount_due_ngn')),
    amount_allocated_ngn: Integer(row.fetch('amount_allocated_ngn')),
    excluded: row['period_status'] == 'excluded' || period.year == Gvac::EXCLUDED_YEAR,
    unresolved: row.fetch('unresolved', false)
  )
end
amount = Integer(payload.fetch('amount_ngn'))
raise ArgumentError, 'payment amount must be positive' unless amount.positive?

allocations = calculator.allocate_oldest_first(
  obligations: obligations,
  payment_amount_ngn: amount
)
proposed = calculator.allocation_coverage(allocations)
as_of = payload.key?('as_of') ? payload.fetch('as_of') : Date.today

updated_obligations = obligations.map do |due|
  added = allocations.sum { |allocated_due, allocated| allocated_due.equal?(due) ? allocated : 0 }
  Gvac::Obligation.new(
    period: due.period,
    amount_due_ngn: due.amount_due_ngn,
    amount_allocated_ngn: due.amount_allocated_ngn + added,
    excluded: due.excluded,
    unresolved: due.unresolved
  )
end
outstanding_after = calculator.outstanding(obligations: updated_obligations, as_of: as_of)

allocation_payload = allocations.map do |due, allocated|
  { period_start: due.period.to_s, amount_allocated_ngn: allocated }
end

coverage = payload['coverage']
if coverage
  start_period = calculator.month(coverage.fetch('start_period'))
  end_period = calculator.month(coverage.fetch('end_period'))
  selected_amount = calculator.coverage_amount(
    obligations: obligations,
    start_period: start_period,
    end_period: end_period
  )
  valid = calculator.coverage_matches_allocation?(
    allocations: allocations,
    start_period: start_period,
    end_period: end_period
  )
else
  start_period = proposed[:start_period]
  end_period = proposed[:end_period]
  selected_amount = proposed[:amount_ngn]
  valid = true
end

puts JSON.generate(
  valid: valid,
  amount_ngn: amount,
  coverage: {
    start_period: start_period.to_s,
    end_period: end_period.to_s,
    amount_ngn: selected_amount
  },
  proposed_coverage: {
    start_period: proposed[:start_period].to_s,
    end_period: proposed[:end_period].to_s,
    amount_ngn: proposed[:amount_ngn]
  },
  allocations: allocation_payload,
  outstanding_after_ngn: outstanding_after
)

File: server/payment_preview.js
const { calendarPeriodStart } = require("./outstanding");

const EXCLUDED_YEAR = 2020;

function month(value) {
  if (typeof value === "string" && /^\d{4}-\d{2}$/.test(value)) {
    return calendarPeriodStart(`${value}-01`);
  }

  return calendarPeriodStart(value);
}

function remaining(due) {
  return Math.max(Number(due.amount_due_ngn) - Number(due.amount_allocated_ngn), 0);
}

function isExcluded(due) {
  const period = month(due.period_start);
  return Boolean(due.excluded) || due.period_status === "excluded" || period.startsWith(`${EXCLUDED_YEAR}-`);
}

function allocateOldestFirst(obligations, paymentAmountNgN) {
  let remainingPayment = paymentAmountNgN;
  const allocations = [];
  const sorted = [...obligations].sort((left, right) => (
    month(left.period_start).localeCompare(month(right.period_start))
  ));

  for (const due of sorted) {
    if (isExcluded(due) || due.unresolved || remaining(due) === 0) continue;

    const required = remaining(due);
    if (remainingPayment < required) break;

    allocations.push({ due, amount_allocated_ngn: required });
    remainingPayment -= required;
  }

  if (remainingPayment !== 0) {
    throw new Error("payment cannot be allocated without a partial payment");
  }

  return allocations;
}

function allocationCoverage(allocations) {
  if (!allocations.length) {
    throw new Error("payment produced no allocations");
  }

  return {
    start_period: month(allocations[0].due.period_start),
    end_period: month(allocations[allocations.length - 1].due.period_start),
    amount_ngn: allocations.reduce((sum, item) => sum + item.amount_allocated_ngn, 0),
  };
}

function coverageMatchesAllocation(allocations, startPeriod, endPeriod) {
  const span = allocationCoverage(allocations);
  return month(startPeriod) === span.start_period && month(endPeriod) === span.end_period;
}

function coverageAmount(obligations, startPeriod, endPeriod) {
  const startDate = month(startPeriod);
  const endDate = month(endPeriod);
  if (startDate > endDate) {
    throw new Error("coverage start must not be after coverage end");
  }

  return obligations.reduce((sum, due) => {
    const period = month(due.period_start);
    if (isExcluded(due) || due.unresolved || period < startDate || period > endDate) {
      return sum;
    }

    return sum + remaining(due);
  }, 0);
}

function outstanding(obligations, asOf) {
  const cutoff = month(asOf);
  return obligations.reduce((sum, due) => {
    if (isExcluded(due) || due.unresolved || month(due.period_start) > cutoff) {
      return sum;
    }

    return sum + remaining(due);
  }, 0);
}

function currentMonthStart() {
  const now = new Date();
  return month(new Date(now.getFullYear(), now.getMonth(), 1));
}

function previewPayment(payload) {
  const amount = Number(payload.amount_ngn);
  if (!Number.isInteger(amount) || amount <= 0) {
    throw new Error("payment amount must be positive");
  }

  const obligations = (payload.obligations || []).map((row) => ({
    period_start: month(row.period_start),
    period_status: row.period_status,
    amount_due_ngn: Number(row.amount_due_ngn),
    amount_allocated_ngn: Number(row.amount_allocated_ngn),
    excluded: row.excluded,
    unresolved: Boolean(row.unresolved),
  }));

  const allocations = allocateOldestFirst(obligations, amount);
  const proposed = allocationCoverage(allocations);
  const asOf = payload.as_of ? month(payload.as_of) : currentMonthStart();

  const updatedObligations = obligations.map((due) => {
    const added = allocations.reduce((sum, item) => (
      item.due === due ? sum + item.amount_allocated_ngn : sum
    ), 0);

    return {
      ...due,
      amount_allocated_ngn: due.amount_allocated_ngn + added,
    };
  });

  const allocationPayload = allocations.map((item) => ({
    period_start: month(item.due.period_start),
    amount_allocated_ngn: item.amount_allocated_ngn,
  }));

  const coverage = payload.coverage;
  let startPeriod;
  let endPeriod;
  let selectedAmount;
  let valid;

  if (coverage) {
    startPeriod = month(coverage.start_period);
    endPeriod = month(coverage.end_period);
    selectedAmount = coverageAmount(obligations, startPeriod, endPeriod);
    valid = coverageMatchesAllocation(allocations, startPeriod, endPeriod);
  } else {
    startPeriod = proposed.start_period;
    endPeriod = proposed.end_period;
    selectedAmount = proposed.amount_ngn;
    valid = true;
  }

  return {
    valid,
    amount_ngn: amount,
    coverage: {
      start_period: startPeriod,
      end_period: endPeriod,
      amount_ngn: selectedAmount,
    },
    proposed_coverage: {
      start_period: proposed.start_period,
      end_period: proposed.end_period,
      amount_ngn: proposed.amount_ngn,
    },
    allocations: allocationPayload,
    outstanding_after_ngn: outstanding(updatedObligations, asOf),
  };
}

module.exports = {
  EXCLUDED_YEAR,
  month,
  remaining,
  allocateOldestFirst,
  allocationCoverage,
  coverageMatchesAllocation,
  coverageAmount,
  outstanding,
  previewPayment,
};
